src/coins.cpp

```diff
@@ -253,11 +253,11 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
 }
 
 bool CCoinsViewCache::Flush() {
-    bool fOk = base->BatchWrite(cacheCoins, hashBlock, /*will_erase=*/true);
+    bool fOk = base->BatchWrite(m_flagged_head.Next(), hashBlock, /*will_erase=*/true);
```

Comment:
Yeah I suppose we can rename it `m_flagged_sentinel` or something? Do you have a suggestion that would have made this clearer if that was the name originally?

Reply:
we have "sentinels" on both ends now, right? A dummy at the beginning and a `nullptr` on the end. This is what I found confusing in the test, why not store the actual head and have a `nullptr` or `sentinel` at the end only? I haven't investigated this, maybe I'm completely off again...

Reply:
> we have "sentinels" on both ends now, right? A dummy at the beginning and a nullptr on the end.

Yes, that sounds right.

> why not store the actual head and have a nullptr or sentinel at the end only?

I found it was more code to keep track of a changing head in the `CCoinsViewCache`, instead of a dummy sentinel I called `m_flagged_head` that remained static and all insertions happened after the sentinel.
If we tracked the actual head, we would have to assign the latest iterator added to be head for each call, and then on deletion we would have to check if it was head as well and then reassign head to the next entry. That last part couldn't be done without access to the `coinsCache`, so we couldn't just add `ClearFlags` into the destructor of the entry.

Reply:
Renamed all `head` to `sentinel`. Hopefully that makes everything clearer :).

Reply:
Thanks, still working on reproducing the benchmarks before I can properly ACK it